File: Python/0_Sort/quickSort.py
from typing import List

class Solution:
    def sortArray(self, nums: List[int]) -> List[int]:
        def quickSort(nums, left, right):
            if left >= right:
                return
            pivot = nums[left + (right - left) // 2]
            l, r = left, right
            while l <= r:
                while nums[l] < pivot:
                    l += 1
                while nums[r] > pivot:
                    r -= 1
                
                if l <= r:
                    nums[l], nums[r] = nums[r], nums[l]
                    l += 1
                    r -= 1
            quickSort(nums, left, r)
            quickSort(nums, l, right)
            
        quickSort(nums, 0, len(nums) - 1)
        return nums

File: CPP/0_Sort/quickSort.cpp
#include <vector>
#include <algorithm>

using namespace std;

/*
Runtime: 485 beat 26%
Memory: 68.96 beat 95%
Time: nlogn
Space: logn (depth of recursion)
*/

class Solution {
public:
    void quickSort(vector<int>& nums, int left, int right) {
        if (left >= right) return;
        int pivot = nums[left + (right - left) / 2];
        int l = left, r = right;
        while (l <= r) {
            while (nums[l] < pivot) l++;
            while (nums[r] > pivot) r--;
            if (l <= r){
                swap(nums[l], nums[r]);
                l++;
                r--;
            }
        }
        quickSort(nums, left, r);
        quickSort(nums, l, right);
    }
    
    vector<int> sortArray(vector<int>& nums) {
        quickSort(nums, 0, nums.size() - 1);
        return nums;
    }
};